ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,72 @@
+
+What makes performance testing different from functional testing is it is harder to decide if the test passed or failed as every performance test tests different aspects of the application and so it expects different performance, different metrics, different thresholds. Furthermore, even if the measured performance of the app under test did not change, there might be some significant resource usage by the application which we want to cause the test to be marked as failed.
+
+The approach to make this pass/fail detection possible proposed here, is to use historical results as a benchmark – not only actual results of the performance test, but also monitoring data about the resource usage, etc.
+
+> Example: Imagine you develop a web e-shop application that uses PostgreSQL data backend. Your perf test is browsing through a goods catalog and measures latency of pages. When you want to decide on pass or fail result of this test, you need to check that metrics like below are aligned to previous results of a same test with same configuration:
+>
+> - Average page load latency.
+> - Backend service CPU and memory usage...
+> - PostgreSQL CPU and memory usage…
+> - Internal PostgreSQL metrics like number of commits during the test or average number of database sessions during the test…
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Checking minimum and maximum of the metric in historical data and making sure the new result is in this range.
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some machine learning thingy.
+
+Goal of this ADR is to propose a way for this kind of testing to be implemented in Stonesoup (feature [STONE-679](https://issues.redhat.com/browse/STONE-679)). Even if it would not deprecate full-blown performance and scale testing, having a release gate with some short and small scale performance test is desirable for many application teams.
+
+We are considering two different options here. They differ in what portion of logic mentioned above is executed where.
+
+### Glossary
+
+- SUT: Subject Under Test
+- Horreum: a results storage service: <https://github.com/Hyperfoil/Horreum>
+- CPT: Continuous Performance Testing
+
+## Decision
+
+Let's use this architecture with a single Horreum instance that would be managed by Red Hat and used by all customers.
```

Comment:
Is this something that would be available to all internal and external tenant clusters?

Reply:
So you are proposing that a single Horreum instance would be used to store, analyze, and output data from both the internal Red Hat tenant cluster and the external public tenant cluster? I feel like this is contrary to the standard. @ralphbean, is there any record of isolation between tenant clusters to be used as a precedence?

Reply:
I assume it is similar to Tekton here: there is one Tekton that is going to run pipelines from both internal teams and external customers? But maybe I'm missing your point.

Reply:
Yeah - on reading here, it makes sense to me to shard this the same way we shard all other services - one instance per member cluster.

Reply:
We are in fact separating Red Hat from other tenants at the cluster level.  Different clusters, different databases.

Reply:
Thanks for separating into one per tenant cluster!